File: app/src/main/java/com/tunjid/rcswitchcontrol/utils/DeletionHandler.java
package com.tunjid.rcswitchcontrol.utils;

import com.google.android.material.snackbar.Snackbar;
import com.tunjid.androidbootstrap.functions.Consumer;

import java.util.Stack;

/**
 * Handles queued deletion of a Switch
 */
public class DeletionHandler<T> extends Snackbar.Callback {

    private final int originalPosition;
    private final Consumer<DeletionHandler<T>> onDismissed;


    private Stack<T> deletedItems = new Stack<>();

   public DeletionHandler(int originalPosition, Consumer<DeletionHandler<T>> onDismissed) {
        this.originalPosition = originalPosition;
        this.onDismissed = onDismissed;
    }

    @Override
    public void onDismissed(Snackbar snackbar, int event) {
        onDismissed.accept(this);
    }

    public boolean hasItems() {
       return !deletedItems.empty();
    }

    public int getDeletedPosition() {
       return originalPosition;
    }

   public void push(T item) {
        deletedItems.push(item);
    }

   public T pop() {
        return deletedItems.pop();
    }
}

File: app/src/main/java/com/tunjid/rcswitchcontrol/utils/DeletionHandler.kt
package com.tunjid.rcswitchcontrol.utils

import com.google.android.material.snackbar.Snackbar
import java.util.*

/**
 * Handles queued deletion of a Switch
 */
class DeletionHandler<T>(
        val deletedPosition: Int,
        private val onDismissed: (self: DeletionHandler<T>) -> Unit) : Snackbar.Callback() {

    private val deletedItems = Stack<T>()

    override fun onDismissed(snackbar: Snackbar?, event: Int) = onDismissed.invoke(this)

    fun hasItems(): Boolean = deletedItems.isNotEmpty()

    fun pop(): T = deletedItems.pop()

    fun peek(): T = deletedItems.peek()

    fun push(item: T) = deletedItems.push(item).let { Unit }
}
